vnext/Shared/Networking/OriginPolicyHttpFilter.cpp

```diff
@@ -0,0 +1,638 @@
+  // This check should be case sensitive.
+  // See also https://fetch.spec.whatwg.org/#http-new-header-syntax
+  if (L"true" != allowCredentials)
+    throw hresult_error{
+        E_INVALIDARG,
+        L"Access-Control-Allow-Credentials value must be \"true\" when the response includes credentials.\\n"};
+};
+
+void OriginPolicyHttpFilter::ValidatePreflightResponse(
+    HttpRequestMessage const &request,
+    HttpResponseMessage const &response) const {
+  auto controlValues = ExtractAccessControlValues(response.Headers());
+
+  auto iRequestArgs = request.Properties().Lookup(L"RequestArgs");
+  // Check if the origin is allowed in conjuction with the withCredentials flag
+  // CORS preflight should always exclude credentials although the subsequent CORS request may include credentials.
+  // if (!CheckAccessStatic(allowedOrigin, allowCredentials, GetOrigin(), withCredentials, /*out*/ errorText))
+  ValidateAllowOrigin(controlValues.AllowedOrigin, controlValues.AllowedCredentials, iRequestArgs);
+
+  // Check if the request method is allowed
+  // if (!IsCrossOriginRequestMethodAllowed(requestMethod, allowedMethodsList, withCredentials, /*out*/ errorText))
+  bool withCredentials = winrt::get_self<RequestArgs, IInspectable>(iRequestArgs)->WithCredentials;
+  bool requestMethodAllowed = false;
+  for (const auto &method : controlValues.AllowedMethods) {
+    if (L"*" == method) {
+      if (!withCredentials) {
+        requestMethodAllowed = true;
+        break;
+      }
+    } else if (boost::iequals(method, request.Method().ToString())) {
+      requestMethodAllowed = true;
+      break;
+    }
+  }
+  if (!requestMethodAllowed)
+    throw hresult_error{
+        E_INVALIDARG,
+        L"Method [" + request.Method().ToString() +
+            L"] is not allowed by Access-Control-Allow-Methods in preflight response.\\n"};
+
+  // Check if request headers are allowed
+  // if (!IsCrossOriginRequestHeadersAllowed(requestHeaders, allowedHeadersList, withCredentials, /*out*/ errorText))
+  // See https://fetch.spec.whatwg.org/#cors-preflight-fetch, section 4.8.7.6-7
+  // Check if the header should be allowed through wildcard, if the request does not have credentials.
+  bool requestHeadersAllowed = false;
+  if (!withCredentials && controlValues.AllowedHeaders.find(L"*") != controlValues.AllowedHeaders.cend()) {
+    // "Authorization" header cannot be allowed through wildcard alone.
+    // "Authorization" is the only member of https://fetch.spec.whatwg.org/#cors-non-wildcard-request-header-name.
+    if (request.Headers().HasKey(L"Authorization") && controlValues.AllowedHeaders.find(L"Authorization") == controlValues.AllowedHeaders.cend())
+      throw hresult_error{
+          E_INVALIDARG,
+          L"Request header field [Authorization] is not allowed by Access-Control-Allow-Headers in preflight response.\\n"};
+
+    requestHeadersAllowed = true;
+  }
+  if (!requestHeadersAllowed) {
```

Comment:
(nit) would you mind adding extra space between end-bracket and following statement? Here and in other places in this file

Reply:
Makes sense. Every if-else block should be in its own "paragraph".